pulsar-io/kinesis/pom.xml

```diff
@@ -38,6 +38,13 @@
       <version>${project.version}</version>
     </dependency>
 
+    <dependency>
+      <groupId>${project.groupId}</groupId>
+      <artifactId>pulsar-functions-instance</artifactId>
+      <version>${project.version}</version>
+      <scope>provided</scope>
```

Comment:
why provided?

Reply:
I have moved `EncryptionContext` out of `Record` (main reason is that it's really tied to Pulsar client API and not relevant to other sources and sinks). 

I have added a new interface `RecordWithEncryptionContext` in `pulsar-functions-instance`, alongside `PulsarSource` implemenentation. 

Sink implementations, can check `s instanceof RecordWithEncryptionContext` and downcast from there to access the `EncryptionContext`. 

The dependency here is "provided" because the jar will always be added by function runtime. In this case we need it to access the interface `RecordWithEncryptionContext`, but there's no need to package it again with the sink.

Reply:
>  I have moved EncryptionContext out of Record (main reason is that it's really tied to Pulsar client API and not relevant to other sources and sinks).

But if Pulsar record is encrypted then this context anyway will be required by any sink. So, I think `EncryptionContext` should be part of `Record`.??

> I have added a new interface RecordWithEncryptionContext in pulsar-functions-instance, 

yes, it seems we should have light weight pulsar-function-common which keeps common types for other modules. so, artifact doesn't have to depend on actual-impl-artifacts such as pulsar-function-instance.

Reply:
Look at https://github.com/apache/incubator-pulsar/pull/2187/files#diff-5df4e50a01720cb0f526ef575cc4d66fR71 for how it got changed for the kinesis sink to access the encryption context. 

More than lightweight, I think it's preferable to separate the API dependencies. Also, that interface really only applies to PulsarRecord and not any other implementations. For the dependency, yes it could be introduced a functions-common, but depending on pulsar-function-instance doesn't add any "real" dependency, since that module is by definition always in the class path, in every execution mode.

Reply:
yes, I saw that change but my only point was that for any Sink-implementation, source will be always a Pulsar and there could be possibility that record is encrypted so, `EncryptionCtx` could be a first class entity so, might be right thing to keep in Record. 
But if you strongly believe to create another type as `RecordWithEncryptionContext` then I will be fine.